docs/src/main/sphinx/develop/supporting-merge.rst

```diff
@@ -0,0 +1,389 @@
+
+The merge case ``RowBlock`` has the following layout:
+
+* Blocks for each column in the table, including partition columns, in
+  table column order.
+* A block containing the matching ``MERGE`` case number matched by the
+  source row, or -1 if no ``MERGE`` case matched the source row.
+* A block containing the ``MERGE`` case operation, encoded as
+  ``INSERT`` = 1, ``DELETE`` = 2, ``UPDATE`` = 3 and if no ``MERGE``
+  case matched, -1.
+
+The page returned from ``transformPage`` consists of all table columns,
+in table column order, followed by the rowId column, followed by the
+operation column from the merge case ``RowBlock``.
+
+Interface ``RowChangeProcessor`` now supports SQL MERGE, and will later be
+used to upgrade the SQL UPDATE engine implementation.
+
+Detecting duplicate matching target rows
+----------------------------------------
+
+The SQL ``MERGE`` specification requires that in each ``MERGE`` case,
+a single target table row must match at most one source row, after
+applying the ``MERGE`` case condition expression.  This is done in
+class ``DuplicateRowFinder.checkForDuplicateTargetRows()``.  That method
+examines successive rows produced by the ``CASE`` expression,
+excluding ``INSERT``ed rows, and checks that the
+``writeRedistributionColumns`` and the ``rowId`` column are the
+identical in successive rows. If so, it raises the
+``MERGE_TARGET_ROW_MULTIPLE_MATCHES`` exception.
+
+``ConnectorMergeTableHandle`` API
+---------------------------------
+
+Interface ``ConnectorMergeTableHandle`` defines one method,
+``getTableHandle()`` to retrieve the ``ConnectorTableHandle``
+originally created by ``ConnectorMetadata.beginMerge()``.
+
+``ConnectorPageSinkProvider`` API
+---------------------------------
+
+To support ``SQL MERGE``, ``ConnectorPageSinkProvider`` must implement
+the method that creates the ``ConnectorMergeSink``:
+
+* ``createMergeSink``::
+
+    ConnectorMergeSink createMergeSink(
+        ConnectorTransactionHandle transactionHandle,
+        ConnectorSession session,
+        ConnectorMergeTableHandle mergeHandle)
+
+  Create a ``ConnectorMergeSink`` for the ``transactionHandle``,
+  ``session`` and ``mergeHandle``.
+
+``ConnectorMergeSink`` API
+--------------------------
+
+As mentioned above, to support ``MERGE``, the connector must define an
+implementation of ``ConnectorMergeSink``, usually layered over the
+connector's ``ConnectorPageSink``.
+
+The ``ConnectorMergeSink`` is created by a call to
+``ConnectorPageSinkProvider.createMergeSink()``.
+
+The only interesting methods are:
+
+* ``storeMergedRows``::
+
+    void storeMergedRows(Page page)
+
+  The Trino engine calls the ``storeMergedRows(Page)`` method of the
+  ``ConnectorMergeSink`` instance returned by
+  ``ConnectorPageSinkProvider.createMergeSink()``, passing the page
+  generated by the ``RowChangeProcessor.transformPage()`` method.
+  That page consists of all table columns, in table column order,
```

Comment:
as above -- All columns? updated columns? does this depend on `io.trino.spi.connector.ConnectorMetadata#supportsMissingColumnsOnInsert`?

Reply:
Yes, all columns.  See earlier comments.